### PR TITLE
`exporters.tflite` initial support

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR adds the `exporters.tflite` subpackage, enabling the export of Transformers models to TFlite.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for your work here. If it's not a blocker, can I review it on Monday?

File: optimum/exporters/tflite/base.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+    def model_to_signatures(
+        self, model: "TFPreTrainedModel", **model_kwargs: Any
+    ) -> Dict[str, "tf.types.experimental.ConcreteFunction"]:
+        """
+        Creates the signatures that will be used when exporting the model to a `tf.SavedModel`.
+        Each signature can be used to perform inference on the model for a given set of inputs.
+
+        Auto-encoder models have only one signature, decoder models can have two, one for the decoder without
+        caching, and one for the decoder with caching, seq2seq models can have three, and so on.
```
Comment (sayakpaul) [inline optimum/exporters/tflite/base.py:271]: * Does this function only exist to aid SavedModel creation? 
* Does `model.save("my_model")` (`model` is an instance of `TFPretrainedModel`) lead to any error? 
* If this helps, TFLite allows us to load specific signatures: https://www.tensorflow.org/lite/guide/signatures

File: optimum/exporters/tflite/convert.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    model.config.return_dict = True
+
+    # Check if we need to override certain configuration item
+    if config.values_override is not None:
+        logger.info(f"Overriding {len(config.values_override)} configuration item(s)")
+        for override_config_key, override_config_value in config.values_override.items():
+            logger.info(f"\t- {override_config_key} -> {override_config_value}")
+            setattr(model.config, override_config_key, override_config_value)
+
+    signatures = config.model_to_signatures(model)
```
Comment (sayakpaul) [inline optimum/exporters/tflite/convert.py:173]: We should be able to call `save()` directly on the model to save it as a `SavedModel` resource unless there's an important reason not to. Ccing @Rocketknight1 @gante. 

Also, do we need to run the intermediate save? I can imagine it's likely there for better standardization of the modules but saving a TF model into SavedModel takes a lot of time (because of the compilation involved). 

Could we directly take a `TFPretrainedModel` or even its concrete function and use that for the TFLite export ([here's](https://colab.research.google.com/gist/sayakpaul/6e4f06232b689afa74a43b1b8e2be243/scratchpad.ipynb) a simple example of obtaining concrete function from a Keras model directly)?

File: optimum/exporters/tflite/convert.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    with TemporaryDirectory() as tmp_dir_name:
+        tf.saved_model.save(model, tmp_dir_name, signatures)
+        converter = tf.lite.TFLiteConverter.from_saved_model(tmp_dir_name)
+        tflite_model = converter.convert()
+
+    with open(output, "wb") as fp:
+        fp.write(tflite_model)
+
+    return config.inputs, config.outputs
```
Comment (sayakpaul) [inline optimum/exporters/tflite/convert.py:183]: There is no optimization involved here. 

TFLite has three post-training optimization offerings through quantization and other things (kernel fusion, for example): https://www.tensorflow.org/lite/performance/post_training_quantization

**Review by sayakpaul — COMMENTED** [review]
Thanks a lot for working on this!

I have left comments on the TFLite export design. Let me know if anything is unclear.

File: optimum/exporters/tflite/base.py (hunk 1 above)
Comment (michaelbenayoun) [reply optimum/exporters/tflite/base.py:271]: - Yes, basically it provides the signatures that will be used for the export of the model.
- I haven't tried, but I think this is equivalent to what we do.
- That's exactly what we use: we export a signature that can be used easily by the `tf.lite.Interpreter`. For decoder and seq2seq models, we will define multiple signatures.

File: optimum/exporters/tflite/convert.py (hunk 3 above)
Comment (michaelbenayoun) [reply optimum/exporters/tflite/convert.py:183]: Yes, of course this is planned. The idea with this PR is to add the minimal code needed to iterate fast later. That is why there is no other configuration that the one for BERT.

File: optimum/exporters/tflite/convert.py (hunk 2 above)
Comment (sayakpaul) [reply optimum/exporters/tflite/convert.py:173]: > I could not find a way to provide signatures in that case

Which case? Could you expand a bit more? 

> It should not provide a speedup in the end because it is (or it more precisely it will be) mandatory to provide a trackable object when exporting with a concrete function, which in turn will do the export to a saved model in the backend.

Could you elaborate a bit more on this? Why providing the trackable object is a requirement?

Comment (michaelbenayoun) [reply optimum/exporters/tflite/convert.py:173]: - I could not find a way to provide signatures when exporting from a concrete function.
- When exporting from a concrete function, if you dont provide a trackable object, it kind of works, but a warning tells you that it will be deprecated => you can also see this warning from your [code sample](https://colab.research.google.com/gist/sayakpaul/6e4f06232b689afa74a43b1b8e2be243/scratchpad.ipynb#scrollTo=KLuOmvf5Fm0B&line=2&uniqifier=1)

Comment (michaelbenayoun) [reply optimum/exporters/tflite/convert.py:173]: Yes, that's not an issue, you can specify the signature you want, that's what I did.

Comment (Rocketknight1) [reply optimum/exporters/tflite/convert.py:173]: Ah, sorry, I was looking at an old version of the page and didn't see the up-to-date conversation!

To support signatures when exporting from a concrete function, you need to make one concrete function per signature. That's exactly what a `ConcreteFunction` is in TensorFlow - it's a function with one precisely defined signature.

Also, if you're getting errors about a missing `trackable_obj`, I'm 99% sure the `trackable_obj` you need is just the original `model` object. [Here](https://colab.research.google.com/gist/Rocketknight1/781023e44fd88f62b67d3c46a7cd50f9/scratchpad.ipynb) is an updated code sample that uses multiple concrete functions and doesn't have the error.

Comment (Rocketknight1) [reply optimum/exporters/tflite/convert.py:173]: Ah, sorry, I follow now! Yeah, I don't know of any good way to name the signatures when using concrete functions.

File: optimum/commands/export/tflite.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+from ...exporters import TasksManager
+
+
+def parse_args_tflite(parser):
+    required_group = parser.add_argument_group("Required arguments")
+    required_group.add_argument(
+        "-m", "--model", type=str, required=True, help="Model ID on huggingface.co or path on disk to load model from."
+    )
+    required_group.add_argument(
+        "output", type=Path, help="Path indicating the directory where to store generated ONNX model."
```
Comment (mfuntowicz) [inline optimum/commands/export/tflite.py:28]: ONNX -> TFlite ?

File: optimum/commands/export/tflite.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        "--sequence_length",
+        type=int,
+        default=None,
+        help=f"Text tasks only. Sequence length {doc_input}",
+    )
+    input_group.add_argument(
+        "--num_choices",
+        type=int,
+        default=None,
+        help=f"Text tasks only. Num choices {doc_input}",
```
Comment (mfuntowicz) [inline optimum/commands/export/tflite.py:80]: Maybe something more detailed? 

What about `Text classification tasks only. How many labels to return` or something like this? wdyt?

File: optimum/commands/export/tflite.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    )
+    input_group.add_argument(
+        "--width",
+        type=int,
+        default=None,
+        help=f"Image tasks only. Width {doc_input}",
```
Comment (mfuntowicz) [inline optimum/commands/export/tflite.py:86]: Image width?

File: optimum/commands/export/tflite.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    )
+    input_group.add_argument(
+        "--height",
+        type=int,
+        default=None,
+        help=f"Image tasks only. Height {doc_input}",
```
Comment (mfuntowicz) [inline optimum/commands/export/tflite.py:92]: Image height?

File: optimum/commands/export/__init__.py [hunk 8]
```diff
@@ -17,12 +17,17 @@
...
 from .. import BaseOptimumCLICommand
 from .onnx import ONNXExportCommand, parse_args_onnx
+from .tflite import TFLiteExportCommand, parse_args_tflite
 
 
 def onnx_export_factory(_):
     return ONNXExportCommand(" ".join(sys.argv[3:]))
 
 
+def tflite_export_factory(_):
```
Comment (fxmarty) [inline optimum/commands/export/__init__.py:27]: Could you add a command line test?

File: optimum/commands/export/tflite.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    )
+    input_group.add_argument(
+        "--num_channels",
+        type=int,
+        default=None,
+        help=f"Image tasks only. Number of channels {doc_input}",
```
Comment (mfuntowicz) [inline optimum/commands/export/tflite.py:98]: Number of channels used to represent the image (RGB = 3, ARGB = 4, GREY =1)?

File: optimum/commands/export/tflite.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    )
+    input_group.add_argument(
+        "--feature_size",
+        type=int,
+        default=None,
+        help=f"Audio tasks only. Feature size {doc_input}",
```
Comment (mfuntowicz) [inline optimum/commands/export/tflite.py:104]: What is feature size here?

File: optimum/commands/export/tflite.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    )
+    optional_group.add_argument("--cache_dir", type=str, default=None, help="Path indicating where to store cache.")
+    optional_group.add_argument(
+        "--trust-remote-code",
+        action="store_true",
+        help="Allow to use custom code for the modeling hosted in the model repository. This option should only be set for repositories you trust and in which you have read the code, as it will execute on your local machine arbitrary code present in the model repository.",
+    )
+
+    input_group = parser.add_argument_group("Input shapes")
+    doc_input = "that the TensorFlow Lite will be able to take as input."
```
Comment (fxmarty) [inline optimum/commands/export/tflite.py:63]:
```suggestion
    doc_input = "that the TensorFlow Lite exported model will be able to take as input."
```

File: optimum/exporters/onnx/config.py [hunk 12]
```diff
@@ -1,5 +1,5 @@
 # coding=utf-8
-# Copyright 2020 The HuggingFace Team. All rights reserved.
+# Copyright 2022 The HuggingFace Team. All rights reserved.
```
Comment (mfuntowicz) [inline optimum/exporters/onnx/config.py:2]: 2023 ? 😄

File: optimum/exporters/tasks.py [hunk 13]
```diff
@@ -1011,6 +1013,8 @@ def get_exporter_config_constructor(
                 The model type to retrieve the config for.
             model_name (`Optional[str]`, defaults to `None`):
                 The name attribute of the model object, only used for the exception message.
+            exporter_config_kwargs( `Optional[Dict[str, Any]]`, defaults to `None`):
```
Comment (fxmarty) [inline optimum/exporters/tasks.py:1016]:
```suggestion
            exporter_config_kwargs(`Optional[Dict[str, Any]]`, defaults to `None`):
```

Comment (fxmarty) [inline optimum/exporters/tasks.py:1016]: just for my knowledge, why is this needed? I see it is only used in `test_tflite_export.py`.

File: optimum/exporters/tflite/__main__.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    model = TasksManager.get_model_from_task(
+        task, args.model, framework="tf", cache_dir=args.cache_dir, trust_remote_code=args.trust_remote_code
+    )
+
+    tflite_config_constructor = TasksManager.get_exporter_config_constructor(model=model, exporter="tflite", task=task)
+    # TODO: find a cleaner way to do this.
+    shapes = {name: getattr(args, name) for name in tflite_config_constructor.func.get_mandatory_axes_for_task(task)}
+    tflite_config = tflite_config_constructor(model.config, **shapes)
+
+    # Ensure the requested opset is sufficient
```
Comment (fxmarty) [inline optimum/exporters/tflite/__main__.py:63]: opset is an ONNX thing no?

File: optimum/exporters/tflite/base.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+    from transformers import PretrainedConfig, TFPreTrainedModel
+
+    from ...utils import DummyInputGenerator
+
+    if is_tf_available():
+        from tensorflow import TensorSpec
+
+
+class MissingMandatoryAxisDimension(ValueError):
+    pass
```
Comment (fxmarty) [inline optimum/exporters/tflite/base.py:40]: should this be in `error_utils.py`?

File: optimum/exporters/onnx/__main__.py [hunk 16]
```diff
@@ -22,17 +22,10 @@
 from ...commands.export.onnx import parse_args_onnx
 from ...utils import DEFAULT_DUMMY_SHAPES, logging
 from ...utils.save_utils import maybe_save_preprocessors
+from ..error_utils import AtolError, OutputMatchError, ShapeError
 from ..tasks import TasksManager
 from .base import OnnxConfigWithPast
-from .convert import (
-    AtolError,
-    OutputMatchError,
-    ShapeError,
```
Comment (fxmarty) [inline optimum/exporters/onnx/__main__.py:30]: Unless mistaken, those three are used in `__main__.py`. Why remove them?

File: optimum/exporters/tflite/base.py [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+            "num_choices": num_choices,
+            "width": width,
+            "height": height,
+            "num_channels": num_channels,
+            "feature_size": feature_size,
+            "nb_max_frames": nb_max_frames,
+            "audio_sequence_length": audio_sequence_length,
+        }
+        for name, value in axes_values.items():
+            setattr(self, name, value)
```
Comment (fxmarty) [inline optimum/exporters/tflite/base.py:136]: This design seem different than the one for ONNX export.

File: optimum/exporters/tflite/config.py [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+logger = logging.get_logger(__name__)
+
+
+class TextEncoderTFliteConfig(TFLiteConfig):
+    """
+    Handles encoder-based text architectures.
+    """
+
+    DUMMY_INPUT_GENERATOR_CLASSES = (DummyTextInputGenerator,)
+    MANDATORY_AXES = ("batch_size", "sequence_length", ("multiple-choice", "num_choices"))
```
Comment (fxmarty) [inline optimum/exporters/tflite/config.py:33]: What is the motivation for `MANDATORY_AXES` compared to the ONNX export?

File: optimum/exporters/tflite/convert.py (hunk 2 above)
Comment (mfuntowicz) [reply optimum/exporters/tflite/convert.py:173]: For reference in `optimum.tir` I was generating on the fly concrete definition with custom signatures, maybe it can help (please dont judge me ...):

```python
    def export_model_to_mlir(self, model: TFPreTrainedModel, examples: Optional = None, dynamic_axes: List[int] = None):
        # Define function signature
        if examples is not None:
            serving_signature = {
                "input_ids": tf.TensorSpec((1, examples[0].shape[1]), dtype=tf.dtypes.int32, name="input_ids"),
                "attention_mask": tf.TensorSpec((1, examples[1].shape[1]), dtype=tf.dtypes.int8, name="attention_mask"),
                # "token_type_ids": tf.TensorSpec((1, None), dtype=tf.dtypes.int8, name="token_type_ids")
            }
        else:
            serving_signature = {
                "input_ids": tf.TensorSpec((1, None), dtype=tf.dtypes.int32, name="input_ids"),
                "attention_mask": tf.TensorSpec((1, None), dtype=tf.dtypes.int8, name="attention_mask"),
                # "token_type_ids": tf.TensorSpec((1, None), dtype=tf.dtypes.int8, name="token_type_ids")
            }

        # Traced function to inject at module-level
        @tf.function(input_signature=[serving_signature])
        def forward(self, inputs):
            outputs = self.eager_serving(inputs)
            return model.serving_output(outputs)

        # Bind the function on the module (TODO: my eyes are already crying blood at this stage ...)
        model.forward = forward.__get__(model, type(model))
```

**Review by mfuntowicz — APPROVED** [review]
LGTM ! Maybe some wording adjustements to be easier for users to grabs some keyword args, but overall it's really cool!

Thanks a lot for bringing this into optimum!

File: optimum/commands/export/tflite.py (hunk 5 above)
Comment (michaelbenayoun) [reply optimum/commands/export/tflite.py:80]: Fixed!

File: optimum/exporters/onnx/config.py (hunk 12 above)
Comment (michaelbenayoun) [reply optimum/exporters/onnx/config.py:2]: Nope, this file was written in 2022!

File: optimum/commands/export/__init__.py (hunk 8 above)
Comment (michaelbenayoun) [reply optimum/commands/export/__init__.py:27]: Done!

File: optimum/exporters/tasks.py (hunk 13 above)
Comment (michaelbenayoun) [reply optimum/exporters/tasks.py:1016]: It is used to provide arguments to the config initalizer when using the `TasksManager`.

File: optimum/exporters/tflite/base.py (hunk 15 above)
Comment (michaelbenayoun) [reply optimum/exporters/tflite/base.py:40]: Not for now. maybe later. but this is quite specific to the `TFLiteConfig`.

File: optimum/exporters/onnx/__main__.py (hunk 16 above)
Comment (michaelbenayoun) [reply optimum/exporters/onnx/__main__.py:30]: I import them from the proper file since it moved.

File: optimum/exporters/tflite/config.py (hunk 18 above)
Comment (michaelbenayoun) [reply optimum/exporters/tflite/config.py:33]: Mandatory axes are the axes which need to be specified to export to `TFLite` since only static shapes are supported (except the batch siize)

Comment (sayakpaul) [reply optimum/exporters/tflite/config.py:33]: Just ensuring we are aware:

https://www.tensorflow.org/api_docs/python/tf/lite/Interpreter#resize_tensor_input

File: optimum/exporters/tasks.py (hunk 13 above)
Comment (fxmarty) [reply optimum/exporters/tasks.py]: Why is it required for tflite and not onnx export?

File: optimum/exporters/tflite/config.py (hunk 18 above)
Comment (michaelbenayoun) [reply optimum/exporters/tflite/config.py:33]: Yes, I tried that once. I do not think it always works. But I would suggest going that path for now and check for dynamic shapes later.

Comment (sayakpaul) [reply optimum/exporters/tflite/config.py:33]: +1 to that. Always better to do it progressively.

File: optimum/exporters/tasks.py (hunk 13 above)
Comment (michaelbenayoun) [reply optimum/exporters/tasks.py]: To provide axes values.
